Bugfix #1080: "Deactivated users reappeared"

chestnutFan opened the pull request:
Fixed #1080 by adding the deactivated member check. Corresponding test functions were also finished.<br>
Details:<br>1. Added the deactivated member check for editing/deleting bills. 
Compare the payer&owers list (POL) of the bill with the active member list (AML). If the relative complement of AML in POL (POL \ AML) is not empty, then the bill contains deactivated members and should not be edited/deleted.<br>2. Added two corresponding integration test functions.<br>
Test:<br>Passed all the tests.

almet commented on ihatemoney/web.py:
```diff
@@ -760,6 +760,19 @@ def delete_bill(bill_id):
+    owers_id = [int(m.id) for m in bill.owers]
+    deactivated_members = set([bill.payer_id] + owers_id).difference(
+        set(active_member_id)
+    )
```
I believe it's possible to ask the model for deactivated users, rather than building it yourself here. You need to look for (`activated=False`).

chestnutFan replied:
Thanks for the suggestion! Will change it.

almet commented on ihatemoney/web.py:
```diff
@@ -776,6 +789,19 @@ def edit_bill(bill_id):
+        flash(
+            _("Deactivated users involved. This bill cannot be edited."),
+            category="warning",
+        )
```
I think having this message shown on a tooltip next to the delete or edit button of the bill would be better, don't you think?

chestnutFan replied:
Good idea. A tooltip makes more sense.

chestnutFan replied:
Hi Almet. I looked into the tooltip thing and found it kind of challenging to me and would take some time to finish. Would you like to keep the flashed message for now?

I'll keep working on it and launch a new pull request once I'm done.

almet replied:
I prefer to merge the pull request once everything is ready :-) Take the time you need and don't hesitate to ask for help. What's blocking you?

chestnutFan replied:
Thanks bro. Just some small issues. Could be done soon.

chestnutFan replied:
Tooltips have been added.

almet reviewed (changes requested):
Thanks for your work on this. It's leading the right way, but needs some changes before integration.

Some feedback : 

1. At all times, avoid copy-pasting code. Here the code to get the deactivated member list is in two different places. If we needed this code, it should probably go in the model instead ;
2. I believe we don't actually need this code because the model can already tel us what are the deactivated members. Try to do this instead of building the list on your own
3. In terms of UX, I believe it would be better to have a tooltip as I explained. I'm happy to change my mind on this if you have good arguments :-)

Anyway, thanks for taking the time to do this :+1:. Let me know when you've made some edits.

chestnutFan commented:
Hi Almet! Thanks for reviewing my code. I rewrote the check function using the activated attribute, and moved it to the model.

chestnutFan commented:
Hi Almet! Tooltips have been added.

almet commented:
This is a bit old now, but I wanted to continue working on it, as I'm currently in the process of tidying a bit this repository.

I've rebased this on top of the latest `main` branch, and it can be merged. But... I'm actually starting to think we're fixing something that is not a problem: if you delete or edit a bill that was involving a deactivated user, I believe we should make it reappear, since its balance is now changed.

Making it so that it's not possible doesn't actually seem right to me, as there are cases where we want to actually change this, for instance if a mistake was made.

So, I'm planning on closing this and #1080 at the same time, unless somebody points me why we should do this.

Cheers!

chestnutFan commented:
> This is a bit old now, but I wanted to continue working on it, as I'm currently in the process of tidying a bit this repository.
> 
> I've rebased this on top of the latest `main` branch, and it can be merged. But... I'm actually starting to think we're fixing something that is not a problem: if you delete or edit a bill that was involving a deactivated user, I believe we should make it reappear, since its balance is now changed.
> 
> Making it so that it's not possible doesn't actually seem right to me, as there are cases where we want to actually change this, for instance if a mistake was made.
> 
> So, I'm planning on closing this and #1080 at the same time, unless somebody points me why we should do this.
> 
> Cheers!

Sounds reasonable to me. Glad to see this one finally been nailed :)
